File: esp32_barrverb/src/BarrVerb.cpp
#include "BarrVerb.h"
#include "rom.h"

void SVF::setFreq(float cutoff, float q, float samplerate) {
    z1 = z2 = 0;
    // Pre-calculate constants
    // w = 2 * tan(pi * cutoff / samplerate)
    // Avoid tan if possible? No, it's init only.
    w = 2.0f * tanf(3.14159f * (cutoff / samplerate));
    a = w / q;
    b = w * w;

    c1 = (a + b) / (1.0f + a * 0.5f + b * 0.25f);
    c2 = b / (a + b);

    d0 = c1 * c2 * 0.25f;
}

float SVF::lpStep(float in) {
    x = in - z1 - z2;
    z2 += c2 * z1;
    z1 += c1 * x;
    return d0 * x + z2;
}

BarrVerb::BarrVerb() {
    // Allocate RAM buffer
    ram = (int16_t*) malloc(sizeof(int16_t) * 16384);
    if (ram) {
        memset(ram, 0, sizeof(int16_t) * 16384);
    }

    // Initialize filters with defaults
    setSampleRate(44100.0f);

    // Default program
    setProgram(0);
}

void BarrVerb::setSampleRate(float sr) {
    sampleRate = sr;
    f1.setFreq(5916.0f, .6572f, sampleRate);
    f2.setFreq(9458.0f, 2.536f, sampleRate);
}

void BarrVerb::setProgram(uint8_t programIndex) {
    program = programIndex;
    if (program > 63) program = 0;
    // Calculate offset in ROM (128 words per program)
    prog_offset = (program & 0x3f) << 7;
}

const char* BarrVerb::getProgramName(uint8_t programIndex) {
    // prog_name is an array of const char* in PROGMEM?
    // In rom.h we defined it as const char* const prog_name[].
    // If the strings themselves are in flash, we might need to handle that,
    // but usually string literals are fine.
    return prog_name[programIndex & 0x3f];
}

void BarrVerb::run(const int16_t *input, int16_t *output, uint32_t frames) {
    if (!ram) return;

    // Load state variables to locals for performance
    int16_t l_acc = acc;
    uint16_t l_ptr = ptr;
    int16_t l_ai = ai;
    int16_t l_li = li;

    // Process 2 frames at a time
    for (uint32_t i = 0; i < frames; i += 2) {
        // --- Input Processing (Sample i) ---
        float inL1 = input[i * 2] / 32768.0f;
        float inR1 = input[i * 2 + 1] / 32768.0f;
        float mono1 = (inL1 + inR1) * 0.5f;
        float lp1 = f2.lpStep(f1.lpStep(mono1));

        // --- Input Processing (Sample i+1) ---
        // We must process the filter for every sample to maintain filter state correctly
        if (i + 1 < frames) {
            float inL2 = input[(i + 1) * 2] / 32768.0f;
            float inR2 = input[(i + 1) * 2 + 1] / 32768.0f;
            float mono2 = (inL2 + inR2) * 0.5f;
            // discard result, just update state.
            // Wait, does the reverb use it? Original code filtered all samples into lowpass[],
            // but then loop for DSP used lowpass[i] where i incremented by 2.
            // So yes, it only used every 2nd filtered sample.
            f2.lpStep(f1.lpStep(mono2));
        }

        // Prepare input for DSP engine
        // Scale factor 2048 matches original code.
        int16_t dsp_in = (int16_t)(lp1 * 2048.0f);

        int16_t out_L = 0;
        int16_t out_R = 0;

        // --- DSP Loop (128 steps) ---
        for (uint8_t step = 0; step < 128; step++) {
            uint16_t opcode = pgm_read_word(&rom[prog_offset + step]);

            switch (opcode & 0xc000) {
                case 0x0000:
                    l_ai = ram[l_ptr];
                    l_li = l_acc + (l_ai >> 1);
                    break;
                case 0x4000:
                    l_ai = ram[l_ptr];
                    l_li = (l_ai >> 1);
                    break;
                case 0x8000:
                    l_ai = l_acc;
                    ram[l_ptr] = l_ai;
                    l_li = l_acc + (l_ai >> 1);
                    break;
                case 0xc000:
                    l_ai = l_acc;
                    ram[l_ptr] = -l_ai;
                    l_li = -(l_ai >> 1);
                    break;
            }

            // Clamp ai to 12-bit range (+/- 2047)
            if (l_ai > 2047) l_ai = 2047;
            else if (l_ai < -2047) l_ai = -2047;

            // Step-specific logic
            if (step == 0x00) {
                // Load RAM with input
                ram[l_ptr] = dsp_in;
            } else if (step == 0x60) {
                // Output Right
                out_R = l_ai;
            } else if (step == 0x70) {
                // Output Left
                out_L = l_ai;
            } else {
                // Latch accumulator
                l_acc = l_li;
            }

            // Update delay pointer
            l_ptr += opcode & 0x3fff;
            l_ptr &= 0x3fff;
        }

        // --- Output Processing ---
        // Scale back up to 16-bit range.
        // Original: (float)ai / 2048. We want int16 (x 32768).
        // 32768 / 2048 = 16.
        int16_t final_L = out_L * 16;
        int16_t final_R = out_R * 16;

        // Write to sample i
        output[i * 2] = final_L;
        output[i * 2 + 1] = final_R;

        // Write to sample i+1 (hold sample)
        if (i + 1 < frames) {
            output[(i + 1) * 2] = final_L;
            output[(i + 1) * 2 + 1] = final_R;
        }
    }

    // Save state variables
    acc = l_acc;
    ptr = l_ptr;
    ai = l_ai;
    li = l_li;
}

File: barrverb-web/src/dsp/BarrVerb.ts
/**
 * BarrVerb DSP Core
 *
 * Ported from https://github.com/ovelhaaa/BarrVerb
 * This implements the SVF and the Instruction Interpreter exactly as in the C++ code.
 */

class SVF {
    private w: number = 0;
    private a: number = 0;
    private b: number = 0;
    private c1: number = 0;
    private c2: number = 0;
    private d0: number = 0;
    private z1: number = 0;
    private z2: number = 0;
    private x: number = 0;

    setFreq(cutoff: number, q: number, samplerate: number) {
        this.z1 = 0;
        this.z2 = 0;
        this.w = 2.0 * Math.tan(Math.PI * (cutoff / samplerate));
        this.a = this.w / q;
        this.b = this.w * this.w;
        this.c1 = (this.a + this.b) / (1.0 + this.a * 0.5 + this.b * 0.25);
        this.c2 = this.b / (this.a + this.b);
        this.d0 = this.c1 * this.c2 * 0.25;
    }

    lpStep(inVal: number): number {
        this.x = inVal - this.z1 - this.z2;
        this.z2 += this.c2 * this.z1;
        this.z1 += this.c1 * this.x;
        return this.d0 * this.x + this.z2;
    }
}

export class BarrVerb {
    private f1: SVF;
    private f2: SVF;

    private ai: number = 0;
    private li: number = 0;
    private acc: number = 0;
    private ptr: number = 0;

    // RAM is 16384 int16 (32KB). In JS we use Int16Array for performance
    private ram: Int16Array;

    // The active program instructions
    private currentProgram: Uint16Array;

    private sampleRate: number = 44100;

    constructor() {
        this.f1 = new SVF();
        this.f2 = new SVF();
        this.ram = new Int16Array(16384);
        this.currentProgram = new Uint16Array(128);
        this.setSampleRate(44100.0);
    }

    setSampleRate(sr: number) {
        this.sampleRate = sr;
        this.f1.setFreq(5916.0, 0.6572, this.sampleRate);
        this.f2.setFreq(9458.0, 2.536, this.sampleRate);
    }

    /**
     * Loads a 128-word program into the cache.
     * @param rom The full ROM Uint16Array
     * @param programIndex The index of the program (0-63)
     */
    setProgram(rom: Uint16Array, programIndex: number) {
        let prog = programIndex;
        if (prog > 63) prog = 0;
        const prog_offset = (prog & 0x3f) << 7; // * 128

        for (let i = 0; i < 128; i++) {
            this.currentProgram[i] = rom[prog_offset + i];
        }
    }

    /**
     * Process a stereo block.
     * inputs/outputs are non-interleaved arrays of Float32.
     * Since this is aimed at Web Audio API AudioWorklet,
     * input[0] is Left channel, input[1] is Right channel.
     * output[0] is Left channel, output[1] is Right channel.
     */
    process(inputL: Float32Array, inputR: Float32Array, outputL: Float32Array, outputR: Float32Array) {
        const frames = inputL.length;

        // Load state to local variables to hint V8 optimizer
        let l_acc = this.acc;
        let l_ptr = this.ptr;
        let l_ai = this.ai;
        let l_li = this.li;
        const l_ram = this.ram;
        const l_prog = this.currentProgram;

        // Original code processes 2 samples at a time in the block:
        // Filter is run every sample, but DSP engine runs every *other* sample.
        // The output of the DSP is written to both sample i and sample i+1.

        for (let i = 0; i < frames; i += 2) {
            // --- Sample i ---
            const mono1 = (inputL[i] + inputR[i]) * 0.5;
            const lp1 = this.f2.lpStep(this.f1.lpStep(mono1));

            // --- Sample i+1 ---
            if (i + 1 < frames) {
                const mono2 = (inputL[i + 1] + inputR[i + 1]) * 0.5;
                // Run filter to update state, but discard output as original does
                this.f2.lpStep(this.f1.lpStep(mono2));
            }

            // Prepare DSP input.
            // Original: dsp_in = (int16_t)(lp1 * 2048.0f)
            let dsp_in = Math.round(lp1 * 2048.0);

            // Clamp strictly to 16-bit integer bounds
            if (dsp_in > 32767) dsp_in = 32767;
            else if (dsp_in < -32768) dsp_in = -32768;

            let out_L = 0;
            let out_R = 0;

            // --- DSP Loop (128 steps) ---
            for (let step = 0; step < 128; step++) {
                const opcode = l_prog[step];

                // Check opcode top 2 bits
                const op = opcode & 0xc000;

                if (op === 0x0000) {
                    l_ai = l_ram[l_ptr];
                    // Using bitwise arithmetic for integer behavior
                    // In JS, >> is a 32-bit signed shift
                    l_li = l_acc + (l_ai >> 1);
                } else if (op === 0x4000) {
                    l_ai = l_ram[l_ptr];
                    l_li = (l_ai >> 1);
                } else if (op === 0x8000) {
                    l_ai = l_acc;
                    l_ram[l_ptr] = l_ai;
                    l_li = l_acc + (l_ai >> 1);
                } else if (op === 0xc000) {
                    l_ai = l_acc;
                    l_ram[l_ptr] = -l_ai;
                    l_li = -(l_ai >> 1);
                }

                // Clamp ai to 12-bit signed range (+/- 2047)
                if (l_ai > 2047) l_ai = 2047;
                else if (l_ai < -2047) l_ai = -2047;

                // Step-specific logic
                if (step === 0x00) {
                    l_ram[l_ptr] = dsp_in;
                } else if (step === 0x60) {
                    out_R = l_ai;
                } else if (step === 0x70) {
                    out_L = l_ai;
                } else {
                    l_acc = l_li;
                }

                // Update delay pointer
                l_ptr += opcode & 0x3fff;
                l_ptr &= 0x3fff;
            }

            // Output scaling
            // Original code: int16_t final_L = out_L * 16;
            // Float output in WebAudio is -1 to 1.
            // 32768 is full 16-bit float scale
            // (out_L * 16) / 32768 = out_L / 2048.0

            const final_L = out_L / 2048.0;
            const final_R = out_R / 2048.0;

            outputL[i] = final_L;
            outputR[i] = final_R;

            if (i + 1 < frames) {
                outputL[i + 1] = final_L;
                outputR[i + 1] = final_R;
            }
        }

        // Save state variables
        this.acc = l_acc;
        this.ptr = l_ptr;
        this.ai = l_ai;
        this.li = l_li;
    }
}
